File: kotlin/src/test/kotlin/tasks/miniaturesUpdate/FurtrackMiniatureUrlResolverTest.kt
package tasks.miniaturesUpdate

import io.ktor.http.*
import testUtils.ExpectedHttpCall
import testUtils.getHttpClientMock
import web.url.FreeUrl
import kotlin.test.*

class FurtrackMiniatureUrlResolverTest{
    @Test
    fun `Test successful retrieval`() {
        val httpClient = getHttpClientMock(
            ExpectedHttpCall(
                "https://orca2.furtrack.com/thumb/49767.jpg",
                null,
                mapOf(),
                "",
                mapOf(),
            ),
            ExpectedHttpCall(
                "https://orca2.furtrack.com/thumb/41933.jpg",
                null,
                mapOf(),
                "",
                mapOf(),
            ),
        )

        val subject = FurtrackMiniatureUrlResolver(httpClient)

        assertEquals(
            "https://orca2.furtrack.com/thumb/49767.jpg",
            subject.getMiniatureUrl(FreeUrl("https://www.furtrack.com/p/49767"))
        )
        assertEquals(
            "https://orca2.furtrack.com/thumb/41933.jpg",
            subject.getMiniatureUrl(FreeUrl("https://www.furtrack.com/p/41933"))
        )
    }

    @Test
    fun `Test non-200 HTTP response`() {
        val httpClient = getHttpClientMock(
            ExpectedHttpCall(
                "https://orca2.furtrack.com/thumb/49767.jpg",
                null,
                mapOf(),
                "",
                mapOf(),
                HttpStatusCode.Forbidden,
            ),
        )

        val subject = FurtrackMiniatureUrlResolver(httpClient)

        val exception = assertFailsWith <MiniatureUrlResolverException> {
            subject.getMiniatureUrl(FreeUrl("https://www.furtrack.com/p/49767"))
        }

        assertEquals(exception.message, "Non-200 HTTP response code")
    }
}

File: symfony/tests/Photos/MiniatureUrlResolver/FurtrackMiniatureUrlResolverTest.php
<?php

declare(strict_types=1);

namespace App\Tests\Photos\MiniatureUrlResolver;

use App\Photos\MiniaturesUpdateException;
use App\Photos\MiniatureUrlResolver\FurtrackMiniatureUrlResolver;
use App\Tests\TestUtils\Http\ExpectedHttpCall;
use App\Tests\TestUtils\Http\HttpClientMockTrait;
use App\Utils\Web\FreeUrl;
use PHPUnit\Framework\Attributes\Small;
use PHPUnit\Framework\TestCase;

#[Small]
class FurtrackMiniatureUrlResolverTest extends TestCase
{
    use HttpClientMockTrait;

    /**
     * @throws MiniaturesUpdateException
     */
    public function testSuccessfulResolve(): void
    {
        $subject = new FurtrackMiniatureUrlResolver(self::getHttpClientMock(
            new ExpectedHttpCall(
                'HEAD',
                'https://orca2.furtrack.com/thumb/49767.jpg',
            ),
            new ExpectedHttpCall(
                'HEAD',
                'https://orca2.furtrack.com/thumb/41933.jpg',
            ),
        ));

        self::assertSame(
            'https://orca2.furtrack.com/thumb/49767.jpg',
            $subject->getMiniatureUrl(new FreeUrl('https://www.furtrack.com/p/49767')),
        );
        self::assertSame(
            'https://orca2.furtrack.com/thumb/41933.jpg',
            $subject->getMiniatureUrl(new FreeUrl('https://www.furtrack.com/p/41933')),
        );
    }

    /**
     * @throws MiniaturesUpdateException
     */
    public function testNon200HttpResponse(): void
    {
        $subject = new FurtrackMiniatureUrlResolver(self::getHttpClientMock(
            new ExpectedHttpCall(
                'HEAD',
                'https://orca2.furtrack.com/thumb/49767.jpg',
                responseCode: 403,
            ),
        ));

        self::expectException(MiniaturesUpdateException::class);
        self::expectExceptionMessage('Non-200 HTTP response code.');
        $subject->getMiniatureUrl(new FreeUrl('https://www.furtrack.com/p/49767'));
    }
}
